File: cn/huohuas001/huhobotPenguin/spigot/qqbind/GroupMsgSender.java
package cn.huohuas001.huhobotPenguin.spigot.qqbind;

import cn.huohuas001.bot.QClient;
import io.github.kloping.qqbot.api.v2.GroupMessageEvent;

/**
 * QQ 群消息发送工具
 *
 * 命令回复（sendWithMention）：
 *   用 event.sendMessage（msg_type=0 纯文本）+ <qqbot-at-user id="openid" /> 标签
 *   QQ 官方机器人 API 在纯文本模式解析此标签为蓝色艾特
 *   参考 At.toString() 新格式：<qqbot-at-user id="..." />
 *   sendMessage 自带 msg_id，实现引用回复效果
 *
 * AI 对话回复（sendReply）：
 *   用 QClient.replyMarkdown（msg_type=2 + markdown + msg_id）
 *   Markdown 模式更好地渲染引用框
 */
public final class GroupMsgSender {
    private GroupMsgSender() {}

    /**
     * 发送带艾特的命令回复（纯文本 + <qqbot-at-user> 标签 + msg_id 引用）。
     */
    public static void sendWithMention(GroupMessageEvent event, String content) {
        try {
            event.sendMessage(content);
        } catch (Throwable t) {
            QqBindManager.logQuiet("[消息发送] 失败: " + t.getMessage());
            try { event.sendMessage(content); } catch (Throwable ignored) {}
        }
    }

    /**
     * 发送 AI 引用回复（Markdown 模式 + msg_id 引用）。
     * replyMarkdown 内部：setMsg_id + setMsg_type(2) + setMarkdown + setContent
     */
    public static void sendReply(GroupMessageEvent event, String content) {
        try {
            QClient.INSTANCE.replyMarkdown(event, content, null);
            QqBindManager.logQuiet("[AI引用] replyMarkdown 发送成功，内容长度=" + (content == null ? 0 : content.length()));
        } catch (Throwable t) {
            QqBindManager.logQuiet("[AI引用] replyMarkdown 失败: " + t.getMessage() + "，回退到 sendMessage");
            try {
                event.sendMessage(content);
                QqBindManager.logQuiet("[AI引用] sendMessage 回退发送成功");
            } catch (Throwable t2) {
                QqBindManager.logQuiet("[AI引用] sendMessage 也失败: " + t2.getMessage());
            }
        }
    }
}

File: common-Bot/src/main/kotlin/cn/huohuas001/huhobotPenguin/spigot/qqbind/GroupMsgSender.kt
package cn.huohuas001.huhobotPenguin.spigot.qqbind

import cn.huohuas001.bot.QClient
import io.github.kloping.qqbot.api.v2.GroupMessageEvent

/**
 * QQ 群消息发送工具。
 *
 * 艾特原理（与 QClient.broadcastGameMessage 一致）：
 * - 艾特格式：`<@openid>`（QQ 不解析 `<qqbot-at-user>`）
 * - 必须用 msg_type=2 + Markdown 模式发送，QQ 才解析 `<@openid>` 艾特；
 * - 纯文本模式（msg_type=0）不解析艾特标签，会原样显示。
 *
 * 引用原理：
 * - QClient.replyMarkdown 内部 setMsg_id(msgId) 实现引用；
 * - Markdown 模式（msg_type=2）渲染引用框。
 *
 * 1.5.3：AI 对话回复改为引用发送者的原消息（QQ 官方 MessageReference 协议，
 * 见 QClient.replyMarkdownWithReference），失败自动回退普通被动回复。
 */
object GroupMsgSender {

    /** 发送带艾特的命令回复（Markdown 模式，失败回退纯文本）。 */
    fun sendWithMention(event: GroupMessageEvent, content: String) {
        try {
            QClient.replyMarkdown(event, content, null)
        } catch (t: Throwable) {
            QqBindManager.logVerbose("[消息发送] replyMarkdown 失败: ${t.message}，回退 sendMessage")
            try {
                event.sendMessage(content)
            } catch (_: Throwable) {
            }
        }
    }

    /**
     * 发送 AI 引用回复（1.5.3：Markdown 模式 + message_reference 引用发送者原消息）。
     *
     * 引用 ID 提取顺序（与官方文档一致）：
     * 新版事件 MessageScene.ext 数组的 msg_idx（REFIDX 形式）→ 事件 msg_id → rawMessage.id；
     * 平台不支持引用协议时自动回退普通被动回复（既有行为），AI 回复不会丢失。
     */
    fun sendReply(event: GroupMessageEvent, content: String) {
        QClient.replyMarkdownWithReference(event, content)
    }
}
